src/terrain/KM_Terrain.pas

```diff
@@ -585,31 +624,189 @@ function TKMTerrain.ScriptTryObjectSet(X, Y: Integer; aObject: Byte): Boolean;
-                UpdateWalkConnect([wcWalk, wcRoad, wcWork], KMRectGrowTopLeft(KMRect(X, Y, X, Y)), DiagonalChanged);
-                Result := True;
-              end;
   end;
+  if aUpdatePassability then
+  begin
+    UpdatePassability(KMRect(X, Y, X, Y)); //When using KMRect map bounds are checked by UpdatePassability
+    UpdateWalkConnect([wcWalk, wcRoad, wcWork], KMRectGrowTopLeft(KMRect(X, Y, X, Y)), aDiagonalChanged);
+  end;
+end;
+
+
+// Try to set an array of Tiles from script. Set terrain, rotation, height and object.
+// Update Passability, WalkConnect and Lighting only once at the end.
+// This is much faster, then set tile by tile with updates on every change
+//
+// Returns True if succeeded
+// use var for aTiles. aTiles can be huge so we do want to make its local copy. Saves a lot of memory
+function TKMTerrain.ScriptTrySetTilesArray(var aTiles: TKMTerrainTileBrief2Array; aRevertOnFail: Boolean; var aErrors: TKMTerrainTileChangeErrorArray): Boolean;
+
+  function IsRectInit(aRect: TKMRect): Boolean;
+  begin
+    Result := (aRect.Left <> -1) and (aRect.Right <> -1)
+          and (aRect.Top <> -1) and (aRect.Bottom <> -1);
+  end;
+
+  procedure UpdateRect(var aRect: TKMRect; X, Y: Integer);
+  begin
+    if not IsRectInit(aRect) then
+      aRect := KMRect(X, Y, X, Y)
+    else
+      KMRectAddPnt(aRect, X, Y);
+  end;
+
+  procedure SetErrorNSetResult(aType: TKMTileChangeType; var aHasErrorOnTile: Boolean; var aErrorType: TKMTileChangeTypeSet; var aResult: Boolean);
+  begin
+    Include(aErrorType, aType);
+    aHasErrorOnTile := True;
+    aResult := False;
+  end;
+
+var I, J, X, Y: Integer;
+    T: TKMTerrainTileBrief;
+    Rect, HeightRect: TKMRect;
+    DiagonalChangedTotal, DiagChanged: Boolean;
+    BackupLand: array of array of TKMTerrainTile;
+    ErrCnt: Integer;
+    HasErrorOnTile: Boolean;
+    ErrorTypesOnTile: TKMTileChangeTypeSet;
+    Error: TKMTerrainTileChangeError;
+begin
+  Result := True;
+  if Length(aTiles) = 0 then Exit;
+
+  //Initialization
+  DiagonalChangedTotal := False;
+  Rect := KMRect(-1, -1, -1, -1);
+  // Use separate HeightRect, because UpdateLight invoked only when Height is changed
+  HeightRect := KMRect(-1, -1, -1, -1);
+  ErrCnt := 0;
+
+  // make backup copy of Land only if we may need revert changes
+  if aRevertOnFail then
+  begin
+    SetLength(BackupLand, fMapY, fMapX);
+    for I := 1 to fMapY do
+      for J := 1 to fMapX do
+        BackupLand[I-1][J-1] := Land[I, J];
+  end;
+
+  for I := 0 to High(aTiles) do
+    for J := 0 to High(aTiles[I]) do
+    begin
+      T := aTiles[I][J];
+      X := I + 1;
+      Y := J + 1;
+
+      HasErrorOnTile := False;
+      ErrorTypesOnTile := [];
+
+      // Update terrain and rotation if needed
+      if tctTerrain in T.ChangeSet then
+      begin
+        if TileInMapCoords(X, Y) and InRange(T.Terrain, 0, 255) and InRange(T.Rotation, 0, 3) then
```

Comment:
Rather than checking TileInMapCoordinates every time, lets just check that High(aTiles) <= fMapY and High(aTiles[I]) <= fMapX before each loop (and only loop to the minimum of these two).

InRange(T.Terrain, 0, 255) is a tautology when your data type is byte ;)